### ISSUE TITLE
Dolphin crashes when object table is full

### ISSUE THREAD

**Issue by fxgallego** [dolphinsmalltalk/Dolphin#194]
In a workspace evaluate:

``` smalltalk
foo := OrderedCollection new.
1 to: 20000000 do: [:each | foo add: Array new]
```

Dolphin crashes without giving any chance to recover from the excessive memory consumption.

**Expected:** An OutOfMemory exception is signaled and the process interrupted before the crash

**Comment by blairmcg** [comment]
Here you are not exhausting memory as such, but rather exceeding the maximum number of objects (which is configurable within finite bounds). Overflowing the maximum object table size is not in general a recoverable situation in Dolphin. If you exceed the object table limit like this, then no further objects can be allocated. The VM detects this and deliberately exits in a controlled way for diagnosis (i.e. fail fast, rather than an uncontrolled crash later). The Smalltalk system needs to allocate objects to do almost anything, including handling a hypothetical "out of memory" signal. One could reserve (i.e. hold back) a block of object table space, but then how much to reserve? No amount of reserve would necessarily be enough to diagnose every situation, and then you just postpone the failure to later and potentially further away from the real problem. This is perhaps one of the unusual cases where debugging the program from within itself is less useful than attaching an external debugger.
In practice I have never found this to be a significant problem myself - on occasion I write some pathological loop and do this sort of thing by accident, but then the crashdump written to the error log is sufficient for diagnosis as it contains a complete stack dump. You also get a specific error code defined in the VMExcept.h file in the VM code:
`#define SE_VMOTFULL			MAKE_CUST_SCODE(SEVERITY_ERROR, FACILITY_NULL, IDP_OTFULL)
`
IDP_OTFULL (i.e. object table full) is 514, which encodes as the Win32 exception code A0000202 you'll see in the crash dump.

In summary, I don't think providing a worthwhile fix for this would be easy, and I'm not at all convinced that the value justifies the effort.

**Comment by fxgallego** [comment]
Blair, sorry, the case is OT full.

We have a problem here in one of our apps when the user executes large queries over the database and in some cases the query could end in an OT full crash. The problem here is that the calculation needed in order to know beforehand the number of Smalltalk objects the query will generate is not easy.

It would be great to have at least a notification about that. We've developed a poors man solution for this in MemoryManager this way
```text
checkOTSize
	^self checkOTSize: self objectCount ! !

checkOTSize: anInteger 
	| count |
	count := self objectCount.
	((anInteger > self otFullSafeLevel and: [count > self otFullSafeLevel]) 
		or: [anInteger > self otFullCriticalLevel]) 
			ifTrue: [OTFullNotification signalWith: (Array with: anInteger with: count)] ! !

otFullSafeLevel
	"Set this to a number below your actual OT size."
	^16000000 ! !

otFullCriticalLevel
	^16500000 ! !

```
And when you are in loops you can send #checkOTSize between loops.
I think that implementing this type of notification is not so hard and gives the oportunity to run some code protected in some way.

**Comment by blairmcg** [comment]
Certainly firing an event when a high water mark in the OT is hit would be feasible.

**Comment by fxgallego** [comment]
Blair, yes that was my first attempt but didn't work very well. I don't remember now what was the problem.  If you want I can check again because maybe is something that can be enhanced in the new VM and finally serves as solution for the problem. We are doing some efforts in order to update our dev branch to D7 and help testing, reporting and probably contributing.
Related to this is the possibility to have a bigger OT, there is no issue for this and I don't want to add one because I don't know if it is feasible.

**Comment by fxgallego** [comment]
Clair, we still has the method you pointed modified this way:
```text
!ProcessorScheduler methodsFor!
 
otOverflow: anInteger
	"Private - The VM has allocated more memory for object headers. Inform the
	memory manager."
 
	MemoryManager current otOverflow: anInteger.
	"This interrupt is only ever sent to the active process, therefore no need to reschedule"
	self enableAsyncEvents: true.
	MemoryManager current checkOTSize: anInteger! !
```
But it appears to not handle in good shape all situations by its own so we have that #checkOTSize that is called when DBRows are being built by the DBConnection framework.
Summarizing, we use both, direct calling a method and the vm callback.

**Comment by blairmcg** [comment]
The otOverflow: notification would always fire, but a subtlety is that it might happen in any running Process. This is one of the reasons that throwing an exception in this case is not a general solution, as it depends where the code is running that is able to handle it. Forwarding the signal to the main process (e.g. by queuing a deferred action) may work for your case, but it really needs to trigger an event that arbitrary observers can listen for and respond to as they see fit.

**Comment by fxgallego** [comment]
I'm testing again our implementation and at least for workspace tests does the job.

In OTFullNotification the defaultAction is:
```text
defaultAction
	self notifyUser.
	self abort
```
It actually aborts any process active while an #otOverflow.
I can send a PR with the changes so you can test this.
What do you think?

**Comment by blairmcg** [comment]
I think the right thing to do in the base system is to trigger an event in MemoryManager - no more. Apps can register for this (e.g. from their SessionManager) and handle it as is appropriate for the application.

**Comment by fxgallego** [comment]
OK yes I agree.

**Comment by blairmcg** [comment]
Finally got around to adding the event in commit linked above
